### PR TITLE
GloVe

### PR DESCRIPTION
**Pull request by ynqa**
### Overview

- Implementation GloVe (on memory)
  - `Solver` is prepared SGD and AdaGrad.

### Feature Works
- GloVe with External Memory (write to disk once before training) like lexvec for huge data.

### CODE REVIEW

File: model/glove/embedding.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+func (s slice) Start() int { return int(s) }
+func (s slice) End() int   { return int(s + 1) }
+func (s slice) Step() int  { return 1 }
+
+// Embedding represents a word embedding. It holds a Tensor, and preslices it for additional performance gains.
+type Embedding struct {
+	P []tensor.Tensor
+	Q []tensor.Tensor
+
+	gradP []tensor.Tensor
```
Comment (chewxy) [inline model/glove/embedding.go]: To think about: is it necessary to keep a copy of the tensors of the gradients?

File: model/glove/glove.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	for i := start; i < end; i++ {
+		p := g.pairs[i]
+		g.train(p.l1, p.l2, p.f)
+	}
+	<-sema
+}
+
+func (g *GloVe) train(pind, qind int, f float64) (err error) {
+	// SGD
+	inner, _ := tensor.Inner(g.emb.P[pind], g.emb.Q[qind])
```
Comment (chewxy) [inline model/glove/glove.go]: bad habit! should always check for errors.

Perhaps create a utility struct somewhere that looks like this:

```
type maybe struct{
    err error
}
func (m *maybe) DoBinOp(fn func(a, b interface{})(tensor.Tensor, error), a, b tensor.Tensor) (retVal tensor.Tensor) {
    if m.err != nil {
        return nil
    }
    retVal, m.err = fn(a, b)
    return
}

func (m *maybe) Error() error { return m.err }
```

then you can do this:

```
m := new(maybe)
inner := m.Do(tensor.Inner, g.emb.p[pind], g.emb.Q[qind])
bias := m.Do(tensor.Add, g.emb.biasP[pind, g.emb.biasQ[qind])
// and so on and so forth
```

File: model/type.go [hunk 3]
```diff
@@ -62,3 +62,8 @@ func (t *Type) RandomTensor(shape ...int) tensor.Tensor {
 	}
 	return ref
 }
+
+// OnesTensor create a tensor has 1 in all elements.
+func (t *Type) OnesTensor(shape ...int) tensor.Tensor {
```
Comment (chewxy) [inline model/type.go]: should be called `Ones`... it already returns a `Tensor`, so no need to repeat

File: model/glove/embedding.go (hunk 1 above)
Comment (ynqa) [reply model/glove/embedding.go]: Yes. GloVe uses AdaGrad as the optimizer, instead of SGD. AdaGrad also use all past iteration's gradients to decrease learning rate.
Here: https://en.wikipedia.org/wiki/Stochastic_gradient_descent#AdaGrad

Comment (chewxy) [reply model/glove/embedding.go]: Yeah I'm familiar with Adagrad. I'm thinking maybe this shouldn't be put into the `Embedding` struct. Instead have something like a `Solver` struct

File: model/type.go (hunk 3 above)
Comment (ynqa) [reply model/type.go]: Sure!

File: model/glove/glove.go (hunk 2 above)
Comment (ynqa) [reply model/glove/glove.go]: Sure. I just wonder how to avoid to write many `if err ...`. it's beneficial info for me, thanks.

File: model/glove/embedding.go (hunk 1 above)
Comment (ynqa) [reply model/glove/embedding.go]: ah I see, that's right. I'll try it.
(It's also good to select other optimizers, `AdaDelta`, `Adam`, and so on.)

File: model/glove/adagrad.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	var diff, cost float64
+	for i := 0; i < a.dimension; i++ {
+		diff += weight[l1+i] * weight[l2+i]
+	}
+	diff += weight[l1+a.dimension] + weight[l2+a.dimension] - f
+	fdiff := diff * coefficient
+	cost = 0.5 * fdiff * diff
+	fdiff *= a.initLearningRate
+
+	for i := 0; i < a.dimension; i++ {
```
Comment (chewxy) [inline model/glove/adagrad.go]: has there any work been done to compare this with a FMA function?

File: model/glove/cofreq.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	"math"
+	"math/rand"
+
+	"github.com/chewxy/lingo/corpus"
+)
+
+// CofreqMap stores the co-frequency between word-word.
+type CofreqMap map[Pair]float64
+
+// Pair stores the co-frequency pair words.
```
Comment (chewxy) [inline model/glove/cofreq.go]: To make this even faster check this out: https://blog.chewxy.com/2017/07/12/21-bits-english/. I'll help you convert when I find the time

File: model/glove/glove.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+	"time"
+
+	"github.com/chewxy/lingo/corpus"
+	"github.com/pkg/errors"
+
+	"github.com/ynqa/word-embedding/model"
+)
+
+// GloVe stores the configs for GloVe models.
+type GloVe struct {
```
Comment (chewxy) [inline model/glove/glove.go]: While GloVe is the proper name (Global Vector) I think it's quite terrible to have random Uppercase letters in the middle of a name that is not a word. Perhaps stick with `Glove` ? Just an opinion

File: model/glove/cofreq.go (hunk 5 above)
Comment (ynqa) [reply model/glove/cofreq.go]: I'll watch it, thanks!

File: model/glove/glove.go (hunk 6 above)
Comment (ynqa) [reply model/glove/glove.go]: Nope, I'll rename it, by the way, why is it terrible there are uppercases in the middle?

File: model/glove/adagrad.go (hunk 4 above)
Comment (ynqa) [reply model/glove/adagrad.go]: Done a little experiments: against about 30 million records
with `FMA` in tensor: 1 min for one iteration
without `FMA` (this): 30 sec for
